librad2/src/keys/types.rs

```diff
@@ -0,0 +1,78 @@
+use std::fmt;
+
+use bs58;
+use sodiumoxide::crypto::sign;
+
+#[derive(Clone, Eq, PartialEq)]
+pub struct DeviceKey(sign::SecretKey);
+pub struct PublicKey(sign::PublicKey);
+pub struct Signature(sign::Signature);
+
+impl DeviceKey {
+    pub fn new() -> Self {
+        let (_, sk) = sign::gen_keypair();
+        DeviceKey(sk)
+    }
+
+    pub fn public_key(&self) -> PublicKey {
+        PublicKey(self.0.public_key())
+    }
+
+    pub fn sign(&self, m: &[u8]) -> Signature {
+        Signature(sign::sign_detached(m, &self.0))
+    }
+}
+
+impl Default for DeviceKey {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl fmt::Display for DeviceKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.public_key().fmt(f)
+    }
+}
+
+impl AsRef<[u8]> for DeviceKey {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+impl From<sign::SecretKey> for DeviceKey {
+    fn from(sk: sign::SecretKey) -> Self {
+        DeviceKey(sk)
+    }
+}
+
+impl PublicKey {
+    pub fn verify(&self, sig: &Signature, m: &[u8]) -> bool {
+        sign::verify_detached(&sig.0, &m, &self.0)
+    }
+}
+
+impl fmt::Display for PublicKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{}",
+            bs58::encode(self.0)
+                .with_alphabet(bs58::alphabet::BITCOIN)
+                .into_string()
+        )
+    }
+}
+
+impl AsRef<[u8]> for PublicKey {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+impl Signature {
+    pub fn verify(&self, m: &[u8], pk: &PublicKey) -> bool {
```

Comment:
> Also, why have `verify` on `PublicKey` _and_ `Signature`?

Because of postfix notation. You can verify either if you also have the other, but depending on which one you have first it may read backwards. I should probably also add a top-level `verify` :trollface: